### PR TITLE
Add username policy alongside the existing password policy.

### PR DESCRIPTION
**Pull request by EreMaijala**
TODO: 

- [x] Consider enabling a default policy.
- [x] Update changelog when merging (default username policy, deprecation of AbstractBase::getCannedPasswordPolicyHint, removal of Database::validateUsernameAndPassword)

### CODE REVIEW

File: module/VuFind/src/VuFind/Auth/AbstractBase.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Auth/AbstractBase.php:325]: Do you think there's any value in keeping this as a deprecated method for backward compatibility, or should we just go ahead and remove it? I don't mind removing it, but if we do, it probably deserves a changelog note, though I imagine local customizations depending on this are pretty rare/unlikely.

File: module/VuFind/src/VuFind/Auth/Database.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Auth/Database.php:207]: Same question here... this might be slightly more likely to be getting called in local custom classes.

File: module/VuFind/src/VuFind/Auth/AbstractBase.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Auth/AbstractBase.php:482]: I wonder if it's worth including a comment like `// e.g. password_minimum_length or username_minimum_length` so that if somebody searches the codebase for those translation strings, they get a hit and don't assume that they are orphaned.

**Review by demiankatz — COMMENTED** [review]
I ran out of time before I could do a full review, but here are some initial comments...

File: config/vufind/config.ini
Comment (demiankatz) [inline config/vufind/config.ini]: Why do we have two examples here? If you intend to have multiple example patterns, we should probably also have multiple corresponding hints.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks again, @EreMaijala -- one more comment.

Also, I tried hands-on testing of this, and when my username failed to match the regex pattern, I got an untranslated "username_error_invalid" message instead of the username_hint I configured. Any idea where that's coming from or why the translation is missing?

(I should note that I tested in a clean environment -- so no old language cache -- and simply uncommented the first username_pattern and username_hint lines in config.ini to test with -- so I didn't do anything particularly weird).

File: module/VuFind/src/VuFind/Auth/AbstractBase.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Auth/AbstractBase.php:325]: Yes, I added it back as a wrapper for `getCannedPolicyHint`.

File: module/VuFind/src/VuFind/Auth/Database.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Auth/Database.php:207]: This is unfortunately not as straightforward, as the split includes the important piece of ignoring username rules when changing password. Without it any previously added user whose username doesn't match current policy would be unable to change password.

File: module/VuFind/src/VuFind/Auth/AbstractBase.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Auth/AbstractBase.php:482]: That's a very valid point. I added comments about the translation strings as well as config settings.

File: config/vufind/config.ini
Comment (EreMaijala) [reply config/vufind/config.ini]: I intended to leave only the second one in there. Now cleaned and explanation added.

**Comment by EreMaijala** [comment]
@demiankatz The translation for `username_error_invalid` was missing. You wouldn't normally see it, but I had also forgotten to enable the length limits and pattern rule on the form after testing without it. So, now you'd get the message only if you managed to avoid the validation on the form.

File: module/VuFind/src/VuFind/Auth/Database.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/Auth/Database.php:207]: I'll add a changelog note.

**Review by demiankatz — CHANGES_REQUESTED** [review]
@EreMaijala, I've gone ahead and updated the changelog on the assumption that this will be merged soon. However, before merging it, I think we need to revise the default username policy, as it currently blocks some characters that I think are quite commonly included in usernames -- e.g. underscores, periods, etc. Additionally, it makes it impossible to use an email address as a username, which is a common preference. Can we continue to block exotic Unicode characters while allowing basic ASCII non-control characters? Maybe that would be a little safer...

**Comment by demiankatz** [comment]
In any case, all tests are passing, so I think this is ready to merge once we agree on what the default username pattern should be!

**Comment by EreMaijala** [comment]
@demiankatz I added some characters to the pattern. Also tests with same pattern. This only leaves a minor issue in that it's not easy to describe clearly what's allowed. Displaying all the characters is messy, so I jsut called it punctuation even though I understand it's not just punctuation.

**Review by demiankatz — APPROVED** [review]
Looks good to me, and all tests are passing. Thanks!

**Pull request merged**
